File: encrypt.js
const fs = require("fs");
const readline = require("readline");
const crypto = require("crypto");

// param constants
const ENCRYPTION_ALGO = "aes-256-cbc";
const SALT_PREFIX = "Salted__"; // standard prefix used by openssl
const KEY_LENGTH = 32;
const PBKDF2_ALGO = "sha256";
const PBKDF2_ROUNDS = 10000;
const ENCRYPTED_CONTENT_ENCODING = "base64";
const UNENCRYPTED_CONTENT_ENCODING = "utf8";

const showHelp = () => {
  console.log("Usage: node encrypt.js [encrypt|decrypt] [options]");
  console.log("");
  console.log("Options:");
  console.log("  --text=<text>           Input text");
  console.log("  --file=<file>           Input file");
  console.log(
    "  --password=<pass>       Password for encryption/decryption. Optional."
  );
  console.log(
    "  --output=<file>         Output file (default: print to stdout)"
  );
  console.log(
    "  --dec-encoding=<enc>    Encoding of the decrypted input/output (base64|utf8) (default: utf8)"
  );
  console.log("  --help                  Show this help message");
};

const parseArgs = () => {
  const args = process.argv.slice(2);
  const action = args[0];
  const options = {};
  for (let i = 1; i < args.length; i++) {
    const arg = args[i];
    if (arg.startsWith("--")) {
      const [key, value] = arg.slice(2).split(/=(.*)/s); // Split only on the first '='
      options[key] = value ? value.trim() : value;
    }
  }
  return { action, options };
};

const getInputContent = (options) => {
  if (options.file)
    return fs.readFileSync(options.file, UNENCRYPTED_CONTENT_ENCODING);
  if (options.text) return options.text;
  return new Promise((resolve) => {
    const rl = readline.createInterface({
      input: process.stdin,
      output: process.stdout,
    });
    rl.question("Enter content: ", (answer) => {
      resolve(answer);
      rl.close();
    });
  });
};

const getPassword = (options) => {
  if (options.password) return options.password;
  return new Promise((resolve) => {
    const rl = readline.createInterface({
      input: process.stdin,
      output: process.stdout,
    });
    rl.question("Enter password: ", (answer) => {
      resolve(answer);
      rl.close();
    });
  });
};

const deriveKeyAndIV = (password, salt) => {
  const derivedKey = crypto.pbkdf2Sync(
    password,
    salt,
    PBKDF2_ROUNDS,
    KEY_LENGTH + 16, // 32 bytes for key + 16 bytes for IV
    PBKDF2_ALGO
  );
  const key = derivedKey.subarray(0, KEY_LENGTH);
  const iv = derivedKey.subarray(KEY_LENGTH);

  return { key, iv };
};

const encrypt = async (content, password) => {
  // Generate salt
  const salt = crypto.randomBytes(8);
  // Derive key and IV from password and salt using PBKDF2
  const { key, iv } = deriveKeyAndIV(password, salt);

  const contentBuffer = Buffer.from(content, UNENCRYPTED_CONTENT_ENCODING);

  const cipher = crypto.createCipheriv(ENCRYPTION_ALGO, key, iv);
  cipher.setAutoPadding(true);
  const encryptedContentBuffer = Buffer.concat([
    cipher.update(contentBuffer),
    cipher.final(),
  ]);

  // the final buffer consist of the salt and the encrypted content
  const finalBuffer = Buffer.concat([
    // add the standard prefix that openssl uses for the salt
    Buffer.from(SALT_PREFIX, "utf8"),
    salt,
    encryptedContentBuffer,
  ]);
  const finalB64 = finalBuffer.toString(ENCRYPTED_CONTENT_ENCODING);

  return finalB64;
};

const decrypt = async (encrypted, password) => {
  const buffer = Buffer.from(encrypted, ENCRYPTED_CONTENT_ENCODING);

  // Obtain parts
  // Omit the prefix when getting the salt.
  const salt = buffer.subarray(SALT_PREFIX.length, 16);
  let encryptedContent = buffer.subarray(16);

  // Derive key and IV from password and salt using PBKDF2
  const { key, iv } = deriveKeyAndIV(password, salt);

  const decipher = crypto.createDecipheriv(ENCRYPTION_ALGO, key, iv);
  // padding
  decipher.setAutoPadding(true);

  const decryptedContent = Buffer.concat([
    decipher.update(encryptedContent),
    decipher.final(),
  ]);

  return decryptedContent.toString(UNENCRYPTED_CONTENT_ENCODING);
};

const main = async () => {
  const { action, options } = parseArgs();
  if (!action) {
    showHelp();
    process.exit(1);
  }

  const content = await getInputContent(options);
  const password = await getPassword(options);

  let output;
  if (action === "encrypt") {
    output = await encrypt(content, password);
  } else if (action === "decrypt") {
    output = await decrypt(content, password, options.decEncoding);
  }

  if (options.output) {
    fs.writeFileSync(options.output, output);
  } else {
    console.log(output);
  }
};

main();

File: encrypt.py
import sys
import getpass
from pathlib import Path
import base64
import secrets
from hashlib import pbkdf2_hmac
from cryptography.hazmat.primitives.ciphers import Cipher, algorithms, modes
from cryptography.hazmat.backends import default_backend

# Constants
ENCRYPTION_ALGO = "aes-256-cbc"
SALT_PREFIX = b"Salted__"
KEY_LENGTH = 32
PBKDF2_ROUNDS = 10000

def show_help():
    print("Usage: python encrypt.py [encrypt|decrypt] [options]")
    print("")
    print("Options:")
    print("  --text=<text>           Input text")
    print("  --file=<file>           Input file")
    print("  --password=<pass>       Password for encryption/decryption")
    print("  --output=<file>         Output file (default: print to stdout)")
    print("  --help                  Show this help message")

def parse_args():
    args = sys.argv[1:]
    if not args or '--help' in args:
        show_help()
        sys.exit(1)
        
    action = args[0]
    options = {}
    
    for arg in args[1:]:
        if arg.startswith('--'):
            if '=' in arg:
                key, value = arg[2:].split('=', 1)
                options[key] = value.strip()
            else:
                options[arg[2:]] = True
                
    return action, options

def derive_key_and_iv(password: str, salt: bytes) -> tuple[bytes, bytes]:
    derived = pbkdf2_hmac(
        'sha256',
        password.encode('utf-8'),
        salt,
        PBKDF2_ROUNDS,
        KEY_LENGTH + 16
    )
    return derived[:KEY_LENGTH], derived[KEY_LENGTH:]

def encrypt(content: str, password: str) -> str:
    salt = secrets.token_bytes(8)
    key, iv = derive_key_and_iv(password, salt)
    
    cipher = Cipher(
        algorithms.AES(key),
        modes.CBC(iv),
        backend=default_backend()
    )
    encryptor = cipher.encryptor()
    
    content_bytes = content.encode('utf-8')
    padding_length = 16 - (len(content_bytes) % 16)
    padded_content = content_bytes + bytes([padding_length] * padding_length)
    
    encrypted_content = encryptor.update(padded_content) + encryptor.finalize()
    final_buffer = SALT_PREFIX + salt + encrypted_content
    
    return base64.b64encode(final_buffer).decode('utf-8')

def decrypt(encrypted: str, password: str) -> str:
    buffer = base64.b64decode(encrypted)
    
    salt = buffer[len(SALT_PREFIX):16]
    encrypted_content = buffer[16:]
    
    key, iv = derive_key_and_iv(password, salt)
    
    cipher = Cipher(
        algorithms.AES(key),
        modes.CBC(iv),
        backend=default_backend()
    )
    decryptor = cipher.decryptor()
    
    decrypted_content = decryptor.update(encrypted_content) + decryptor.finalize()
    padding_length = decrypted_content[-1]
    unpadded_content = decrypted_content[:-padding_length]
    
    return unpadded_content.decode('utf-8')

def get_input_content(options: dict) -> str:
    if 'file' in options:
        return Path(options['file']).read_text()
    if 'text' in options:
        return options['text']
    return input("Enter content: ")

def get_password(options: dict) -> str:
    if 'password' in options:
        return options['password']
    return getpass.getpass("Enter password: ")

def main():
    action, options = parse_args()
    
    content = get_input_content(options)
    password = get_password(options)
    
    if action == 'encrypt':
        output = encrypt(content, password)
    elif action == 'decrypt':
        output = decrypt(content, password)
    else:
        show_help()
        sys.exit(1)
    
    if 'output' in options:
        Path(options['output']).write_text(output)
    else:
        print(output)

if __name__ == '__main__':
    main()